hivemind/client/averaging/load_balancing.py

```diff
@@ -0,0 +1,95 @@
+        throughputs = [throughput if throughput is not None else default_throughput for throughput in throughputs]
+        fractions = optimize_parts_lp(vector_size, np.asarray(throughputs), min_size)
+    else:
+        assert not all(throughput == 0 for throughput in throughputs), "Must have at least one nonzero throughput"
+        fractions = np.asarray([1.0 if throughput is None else 0.0 for throughput in throughputs])
+
+    return tuple(hagenbach_bishoff(vector_size, fractions))
+
+
+def optimize_parts_lp(vector_size: int, throughputs: np.ndarray, min_size: int = 0, eps: float = 1e-15) -> np.ndarray:
+    """
+    This method solves an optimization problem to minimize the total allreduce time.
+    In butterfly all-reduce, each peer acts both as a "client" and as an "aggregator":
+    * a "client" splits his local vector into shards and sends each shard to one peer, then downloads the average
+    * an "aggregator" receives a certain part of vector components from all peers, aggregates and returns the average
+
+    Peer i network load as a "client" = vector_size * (1 - fraction_assigned_to_peer_i)
+    Peer i network load as an "aggregator" = vector_size * (group_size - 1) * fraction_assigned_to_peer_i
+    Peer i total communication = vector_size * [1 + (group_size - 2) * fraction_assigned_to_peer_i]
+    Total time = max_i (total_communication_for_peer_i / throughputs[i])
+
+    We find optimal vector fractions by minimizing the total time (using https://tinyurl.com/minimax-to-lp )
+    Then, we use Hagenbach-Bishoff apportionment to split the finite vector based on the optimal fractions.
+
+    """
+    assert np.all(throughputs >= 0) and np.any(throughputs > 0)
+    permutation = np.argsort(-throughputs)
+    throughputs = throughputs[permutation]
+    is_nonzero = throughputs != 0
+
+    group_size = len(throughputs)
+    num_variables = group_size + 1  # [w_1, ..., w_N, ksi]
+
+    c = np.zeros(num_variables)
+    c[-1] = 1.0  # optimize w.r.t. ksi
+
+    # the constraints below are tuples (A, b) such that Ax <= b
+    nonnegative_weights = -np.eye(group_size, M=num_variables), np.zeros(group_size)
+    weights_sum_to_one = c[None, :] - 1.0, np.array([-1.0])
+    coeff_per_variable = (group_size - 2.0) / np.maximum(throughputs, eps)
+    coeff_matrix_minus_ksi = np.hstack([np.diag(coeff_per_variable), -np.ones((group_size, 1))])
+    ksi_is_maximum = coeff_matrix_minus_ksi[is_nonzero], -1.0 / throughputs[is_nonzero]
+    force_max_weights = np.eye(group_size, M=num_variables), is_nonzero.astype(c.dtype)
+
+    A, b = list(map(np.concatenate, zip(nonnegative_weights, weights_sum_to_one, ksi_is_maximum, force_max_weights)))
+
+    solution = scipy.optimize.linprog(c, A_ub=A, b_ub=b)
+    if solution.success:
+        peer_fractions = solution.x[:group_size]
+        if np.max(peer_fractions) >= min_size / float(vector_size):
+            peer_fractions[peer_fractions < min_size / float(vector_size)] = 0.0
```

Comment:
Do the fractions still sum to 1 in that case?

Reply:
renormalized on the next line